sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java

```diff
@@ -0,0 +1,517 @@
+    }
+
+    /**
+     * Returns the size of the table in bytes, does not take into consideration
+     * filter/partition details passed, if any.
+     */
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+      Configuration conf = new Configuration();
+      for (Entry<String, String> entry : spec.getConfigProperties().entrySet()) {
+        conf.set(entry.getKey(), entry.getValue());
+      }
+      IMetaStoreClient client = null;
+      try {
+        HiveConf hiveConf = HCatUtil.getHiveConf(conf);
+        client = HCatUtil.getHiveMetastoreClient(hiveConf);
+        Table table = HCatUtil.getTable(client, spec.getDatabase(),
+              spec.getTable());
+        return StatsUtils.getFileSizeForTable(hiveConf, table);
+      } finally {
+        //IMetaStoreClient is not AutoCloseable, closing it manually
+        client.close();
+      }
+    }
+
+    /**
+     * Calculates the 'desired' number of splits based on desiredBundleSizeBytes which is passed as
+     * a hint to native API.
+     * Retrieves the actual splits generated by native API, which could be different from the
+     * 'desired' split count calculated using desiredBundleSizeBytes
+     * @throws HCatException
+     */
+    @Override
+    public List<BoundedSource<DefaultHCatRecord>> split(long desiredBundleSizeBytes,
+                                                PipelineOptions options) throws Exception {
+      LOG.debug("desiredBundleSize {} bytes", desiredBundleSizeBytes);
+      List<BoundedSource<DefaultHCatRecord>> sources = new ArrayList<>();
+      int desiredSplitCount = 1;
+      long estimatedSizeBytes = getEstimatedSizeBytes(options);
+      if (desiredBundleSizeBytes > 0 && estimatedSizeBytes > 0) {
+        desiredSplitCount = (int) Math.ceil((double) estimatedSizeBytes / desiredBundleSizeBytes);
+      }
+      LOG.debug("desiredSplitCount {}", desiredSplitCount);
+      ReaderContext readerContext = getReaderContext(desiredSplitCount);
+      //process the splits returned by native API
+      //this could be different from 'desiredSplitCount' calculated above
+      LOG.debug("actual split count {}", readerContext.numSplits());
+      for (int split = 0; split < readerContext.numSplits(); split++) {
+        sources.add(new BoundedHCatalogSource(spec.withSplitId(split).withContext(readerContext)));
+      }
+      return sources;
+    }
+
+    private ReaderContext getReaderContext(long desiredSplitCount) throws HCatException {
+      ReadEntity entity = new ReadEntity.Builder().withDatabase(spec.getDatabase())
+          .withTable(spec.getTable()).withFilter(spec.getFilter()).build();
+      Map<String, String> configProps = spec.getConfigProperties();
+      //pass the 'desired' split count as an hint to the API
+      configProps.put(HCatConstants.HCAT_DESIRED_PARTITION_NUM_SPLITS,
+          String.valueOf(desiredSplitCount));
+      configProps = Collections.unmodifiableMap(configProps);
```

Comment:
Agreed, rather than making a copy here I made this while building the pipeline - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java#L172